onnxruntime/core/providers/tvm/tvm_so_execution_provider.cc

```diff
@@ -0,0 +1,301 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include <fstream>
+#include <streambuf>
+#include <map>
+
+#include "core/framework/execution_provider.h"
+#include "core/framework/tensorprotoutils.h"
+#include "core/framework/kernel_registry.h"
+#include "core/framework/compute_capability.h"
+#include "core/platform/env.h"
+#include "core/graph/model.h"
+
+#include "tvm_so_execution_provider.h"
+#include "xpu_data_transfer.h"
+#include "tvm_allocator.h"
+#include "tvm_utils.h"
+#include "tvm_api.h"
+#include "hash_alg/hasher.h"
+
+
+using namespace ONNX_NAMESPACE;
+
+namespace onnxruntime {
+namespace tvm {
+
+// Information to construct kernel function state.
+struct TVMFuncState {
+  AllocateFunc allocate_func = nullptr;
+  DestroyFunc release_func = nullptr;
+  AllocatorHandle allocator = nullptr;
+  std::shared_ptr<TVMCompilerBase> compiler = nullptr;
+};
+
+TvmSoExecutionProvider::TvmSoExecutionProvider(const TvmEPOptions& options)
+    : IExecutionProvider{kTvmExecutionProvider},
+      options_{options} {
+  AllocatorCreationInfo default_memory_info = {[](int) {
+                                                 return std::make_unique<TVMAllocator>();
+                                               },
+                                               0, false};
+  allocator_ = CreateAllocator(default_memory_info);
+  InsertAllocator(allocator_);
+
+  // Get environment variables
+  const Env& env_instance = Env::Default();
+
+  const std::string dump_subgraphs_env = env_instance.GetEnvironmentVar(env_vars::kDumpSubgraphs);
+  if (!dump_subgraphs_env.empty()) {
+    LOGS(*GetLogger(), INFO) << "TVM EP processing shared lib does not save subgraphs " <<
```

Comment:
Does it support subgraphs? Maybe it is better to say it that way.

Reply:
Thanks! I've corrected it.